baja/src/baja_session.rs

```diff
@@ -9,38 +11,136 @@
 //! See the `LICENSE.markdown` file in the Veracruz root directory for
 //! information on licensing and copyright.
 
-use super::BajaError;
-use rustls::Session;
-use std::io::{Read, Write};
-use std::vec::Vec;
-use veracruz_utils;
+use crate::error::BajaError;
+use veracruz_utils::VeracruzRole;
+
+use std::{
+    io::{Read, Write},
+    vec::Vec,
+};
+
+use rustls::{Certificate, ServerSession, Session};
+
+////////////////////////////////////////////////////////////////////////////////
+// Principals.
+////////////////////////////////////////////////////////////////////////////////
+
+/// A principal is an individual identified with a cryptographic certificate,
+/// and assigned a set of roles that dictate what that principal can and cannot
+/// do in a Veracruz computation.
+pub struct Principal {
```

Comment:
This `Principal` type is duplicated with `VeracruzIdentity` in veracruz_util. One trick that might work is `type Principal = VeracruzIdentity` but I am not sure if we can add more methods to the `Principal` type.

Reply:
Good point.  I'll see whether there's some unification that can be done, here.

Reply:
I managed to unify the two types by introducing a type-variable to abstract over the different representations of cryptographic certificates that the two used, i.e. either `String` or the `Certificate` type from the `RusTLS` library, and moved extra material across into `veracruz-utils`.